### PR TITLE
[FIX] base: prevent selecting unsupported "action types" in scheduled actions

### PR DESCRIPTION
**Pull request by jhk-odoo**
Currently, scheduled Actions only work if you choose "Python Code" as the action type ("Action To Do"). Other action types are not really supported as they work "record-wise" (using `active_ids`) which makes no sense for scheduled actions.

We should help users avoid surprises by hiding the "Action To Do" field on scheduled actions, and default to "Python Code" implicitly.

----

Example steps to reproduce :
- install invoicing
- go to scheduled actions > create a new action > select invoicing as model,
select "Execute several actions" as action to do > save
- edit > add a new action in the actions tab > create > create an action that
executes python code and add the line below
log('test logging', level='info')
- save and run your scheduled action manually
- go to settings > technical > logging
- nothing has been logged

opw-2227082

### CODE REVIEW

**Comment by jhk-odoo** [comment]
@nle-odoo

**Comment by nim-odoo** [comment]
Is it because of this:

https://github.com/odoo/odoo/blob/d8d60eef3fc412aa7a314ae56f36d60437185508/odoo/addons/base/models/ir_actions.py#L544

(`action.state` can be equal to `multi`)

If so, some extra comment in the diff would be nice, because the reason is not obvious.

**Comment by nle-odoo** [comment]
yeah, the `run_action_%s_multi` action is always called one time (when available for a type of action), but since there is no `run_action_multi_multi` instead we call `run_action_multi` for each "active_ids".

so if we had an action:

- actionA: multi, composed of:
  - actionB: code
  - actionC: object_write
  - actionD: multi, composed of:
    - actionE: code
    - actionF: object_write

if we ran it with current code with 0 records, we would execute no action, with 2 record we would execute:

- actionA(active_id=record(1))
- actionB(active_id=record(1))
- actionC(active_id=record(1))
- actionD(active_id=record(1))
- actionE(active_id=record(1))
- actionF(active_id=record(1))
- actionA(active_id=record(2))
- actionB(active_id=record(2))
- actionC(active_id=record(2))
- actionD(active_id=record(2))
- actionE(active_id=record(2))
- actionF(active_id=record(2))

with the current change on this PR, with 0 records we would execute:

- actionA(active_id=record())
- actionB(active_id=record())
- actionD(active_id=record())
- actionE(active_id=record())

and with 2 records we would execute:

- actionA(active_id=record(1,2))
- actionB(active_id=record(1,2))
- actionC(active_id=record(1))
- actionC(active_id=record(2))
- actionD(active_id=record(1,2))
- actionE(active_id=record(1,2))
- actionF(active_id=record(1))
- actionF(active_id=record(2))

**Comment by nle-odoo** [comment]
Hi @odony 


what do you think about this? should this be done in stable or master, this has the risk that:

- if a `multi` action contained `code` action, before when you ran it without active_id => nothing was done => now the code will be executed

- if a `multi` action contained `code` action before when you ran it with several active_ids => the code would be executed once per active_id => now the code would be executed once (like "code" action when they are not wrapped in `multi`)

**Comment by odony** [comment]
> what do you think about this? should this be done in stable or master, this has the risk that:
> 
> * if a `multi` action contained `code` action, before when you ran it without active_id => nothing was done => now the code will be executed once
> * if a `multi` action contained `code` action before when you ran it with several active_ids => the code would be executed once per active_id => now the code would be executed once (like "code" action when they are not wrapped in `multi`)

The second part seems like it will impact _automated_ actions and break them if they are not written in a way that supports several `active_ids`. That doesn't look like a good idea in stable.

Maybe it would be best to just forbid "multi" type for scheduled actions, as it's not something that was intended to work in the first place. I suspect the only type of action that was supposed to work for "scheduled actions" is "code", and that's it.

Most other actions don't make sense and do not work if there is no active_id (i.e. in "scheduled action mode"):
- write object
- add followers
- send email
- send sms
- create next activity
- ...

We can ask @tde-banana-odoo as he authored 1c79c5ee3050ad8ba907aeda70c0d1f3dbe837a0 but I suspect we could just hide the `state` field when `usage == 'ir_cron'`, and default to `code` all the time.

**Comment by odony** [comment]
I did not mention it but of course the "multi" feature is kind of useless for scheduled actions anyway as you can simply create several actions - all those things were designed for automated actions initially, and scheduled actions inherited them somewhat accidentally.

**Comment by nle-odoo** [comment]
Thanks.

The behavior is exactly the same on normal server action with "Run" button (that appear when the action model_id is ir.actions.server), so if we don't change the behavior of multi I think it would also be good to hide the "Run" button on ir.actions.server if the `state='multi'`.

For scheduled action, for stable, we could have `{'default_state': 'code'}` in window action context and maybe have `{'readonly': [['state', '=', 'code']]}` to prevent changing it for new scheduled action  but allow to modify the existing ones to `code`.

And for multi, there is only code that does not need `active_id` (there is also multi action for sms and followers but they are called with a recordset containing active_ids)

**Comment by odony** [comment]
> The behavior is exactly the same on normal server action with "Run" button (that appear when the action model_id is ir.actions.server), so if we don't change the behavior of multi I think it would also be good to also hide the "Run" button on ir.actions.server if `state != 'code'`.

Sure, makes sense.


> For scheduled action, for stable, we could have `{'default_state': 'code'}` in window action context and maybe have `{'readonly': [['state', '=', 'code']]}` to prevent changing it for other scheduled action type but allow to modify the existing ones to `code`.

Why not have the default in the Python code, but only in the scheduled action model, so that no action is required for users to get the fix?
As for the `state` field I would just remove it from the view for scheduled actions entirely, to avoid a weird behavior that is too difficult to understand for users. Nobody should have a scheduled action that is not `code` anyway, as they simply don't work, so there's no big loss - they can be deleted and re-created.

Let's confirm this with @tde-banana-odoo and @rco-odoo perhaps?

**Comment by nim-odoo** [comment]
ping @rco-odoo @tde-banana-odoo

**Comment by tde-banana-odoo** [comment]
Between code, commit message and discussions that are not really straightforward to understand I have to admit I don't understand what you really expect from me.

**Comment by nle-odoo** [comment]
tl;dr

Current proposition of change is:

- on ir.actions.server: hide "Run" button when the action type is different than "Code" (it does nothing for all other type of server action)

- on base.automation: have "code" ("Execute Python Code") by default as state (it's the only one that work for a base.automation since all other state needs active_ids/active_id) and hide the "Action To Do" field on the form view.

> odony: Let's confirm this with @tde-banana-odoo and @rco-odoo perhaps?

**Comment by odony** [comment]
@tde-banana-odoo I've rewritten the PR title and desc, might help clarify the idea too

```
[FIX] base: prevent selecting unsupported "action types" in scheduled actions

Currently, scheduled Actions only work if you choose "Python Code" as the action type ("Action To Do"). Other action types are not really supported as they work "record-wise" (using `active_ids`), which makes no sense for scheduled actions.

We should help users avoid surprises by hiding the "Action To Do" field on scheduled actions, and default to "Python Code" implicitly.
```

**Comment by odony** [comment]
Reminder: it should be noted that all scheduled actions included in standard are of "code" type, and if anyone has created a non-code scheduled action, it does not work.

File: odoo/addons/base/views/ir_actions_views.xml [hunk 1]
```diff
@@ -264,7 +264,7 @@
                                 help="Remove 'More' top-menu contextual action related to this action"/>
                         <button name="run" string="Run" type="object"
                                 class="btn-primary"
-                                attrs="{'invisible':[('model_id', '!=', %(base.model_ir_actions_server)s)]}"
+                                attrs="{'invisible':[('model_id', '!=', %(base.model_ir_actions_server)s), ('state', '!=', 'code')]}"
```
Comment (nle-odoo) [inline odoo/addons/base/views/ir_actions_views.xml:267]: here it should be `'|'` before the 2 leaves

**Comment by jhk-odoo** [comment]
@nle-odoo updated

**Comment by nim-odoo** [comment]
@nle-odoo ping

**Comment by nle-odoo** [comment]
robodoo r+
